### PR TITLE
Doc base macros

### PR DESCRIPTION
**Pull request by matbesancon**
A simple PR adding missing documentation to some macros.

Some I couln't find anything to document (or couldn't get the code), like `@irrational`.

### CODE REVIEW

**Comment by mbesancon** [comment]
I think re-launching Travis will be necessary for the build which failed:
```
E: There were unauthenticated packages and -y was used without --allow-unauthenticated
apt-get.diagnostics
apt-get install failed
$ cat ~/apt-get-update.log
Ign:1 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty InRelease
Get:2 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates InRelease [65.9 kB]
Get:3 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-backports InRelease [65.9 kB]
Hit:4 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty Release
Ign:5 http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.4 InRelease
Get:7 http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.4 Release [2,495 B]
Get:9 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/main Sources [515 kB]
Get:10 http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.4 Release.gpg [801 B]
Get:11 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/universe Sources [254 kB]
Get:12 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/multiverse Sources [7,226 B]
Get:13 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/main amd64 Packages [1,353 kB]
Get:14 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/main i386 Packages [1,278 kB]
Ign:8 http://toolbelt.heroku.com/ubuntu ./ InRelease
Get:15 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/main Translation-en [651 kB]
Get:17 http://security.ubuntu.com/ubuntu trusty-security InRelease [65.9 kB]
Get:18 http://ppa.launchpad.net/chris-lea/redis-server/ubuntu trusty InRelease [15.4 kB]
Get:19 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/restricted amd64 Packages [21.4 kB]
Get:20 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/restricted i386 Packages [21.1 kB]
Get:21 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/restricted Translation-en [3,704 B]
Get:22 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/universe amd64 Packages [592 kB]
Get:23 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/universe i386 Packages [581 kB]
Get:24 https://download.docker.com/linux/ubuntu trusty InRelease [26.5 kB]
Get:25 https://dl.hhvm.com/ubuntu trusty InRelease [2,411 B]
Hit:16 http://toolbelt.heroku.com/ubuntu ./ Release
Get:26 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/universe Translation-en [303 kB]
Get:28 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/multiverse amd64 Packages [16.0 kB]
Get:29 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/multiverse i386 Packages [16.5 kB]
Get:30 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-updates/multiverse Translation-en [7,680 B]
Ign:31 http://dl.google.com/linux/chrome/deb stable InRelease
Get:32 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-backports/universe Sources [41.3 kB]
Get:33 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-backports/universe amd64 Packages [52.5 kB]
Get:34 http://us-central1.gce.archive.ubuntu.com/ubuntu trusty-backports/universe i386 Packages [52.4 kB]
Get:35 http://apt.postgresql.org/pub/repos/apt trusty-pgdg InRelease [61.4 kB]
Err:10 http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.4 Release.gpg
  The following signatures were invalid: KEYEXPIRED 1515625755
Get:36 http://dl.google.com/linux/chrome/deb stable Release [1,189 B]
Get:6 http://dl.bintray.com/apache/cassandra 39x InRelease [3,168 B]
Get:37 https://dl.hhvm.com/ubuntu trusty/main amd64 Packages [1,803 B]
Get:39 https://download.docker.com/linux/ubuntu trusty/stable amd64 Packages [3,323 B]
Get:40 https://download.docker.com/linux/ubuntu trusty/edge amd64 Packages [4,527 B]
Ign:41 http://ppa.launchpad.net/couchdb/stable/ubuntu trusty InRelease
Get:42 http://dl.google.com/linux/chrome/deb stable Release.gpg [819 B]
Get:43 http://security.ubuntu.com/ubuntu trusty-security/main Sources [201 kB]
Get:44 http://ppa.launchpad.net/git-core/ppa/ubuntu trusty InRelease [15.4 kB]
Get:45 http://dl.google.com/linux/chrome/deb stable/main amd64 Packages [1,119 B]
Hit:46 https://packagecloud.io/computology/apt-backport/ubuntu trusty InRelease
Get:47 http://apt.postgresql.org/pub/repos/apt trusty-pgdg/main amd64 Packages [186 kB]
Get:48 https://packagecloud.io/github/git-lfs/ubuntu trusty InRelease [23.2 kB]
Get:49 http://ppa.launchpad.net/openjdk-r/ppa/ubuntu trusty InRelease [15.4 kB]
Get:50 http://security.ubuntu.com/ubuntu trusty-security/restricted Sources [5,050 B]
Get:51 https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu trusty InRelease [23.7 kB]
Get:52 http://security.ubuntu.com/ubuntu trusty-security/universe Sources [89.7 kB]
Get:53 http://security.ubuntu.com/ubuntu trusty-security/multiverse Sources [3,010 B]
Get:54 https://packagecloud.io/github/git-lfs/ubuntu trusty/main amd64 Packages [6,490 B]
Get:55 http://apt.postgresql.org/pub/repos/apt trusty-pgdg/main i386 Packages [185 kB]
Get:56 http://ppa.launchpad.net/pollinate/ppa/ubuntu trusty InRelease [15.4 kB]
Get:57 http://security.ubuntu.com/ubuntu trusty-security/main amd64 Packages [929 kB]
Get:58 https://packagecloud.io/github/git-lfs/ubuntu trusty/main i386 Packages [6,260 B]
Get:59 https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu trusty/main amd64 Packages [5,515 B]
Get:60 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty InRelease [15.4 kB]
Get:61 https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu trusty/main i386 Packages [5,515 B]
Get:62 http://security.ubuntu.com/ubuntu trusty-security/main i386 Packages [856 kB]
Ign:60 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty InRelease
Get:63 http://security.ubuntu.com/ubuntu trusty-security/main Translation-en [493 kB]
Get:64 http://ppa.launchpad.net/webupd8team/java/ubuntu trusty InRelease [15.5 kB]
Get:65 http://security.ubuntu.com/ubuntu trusty-security/restricted amd64 Packages [18.1 kB]
Get:66 http://security.ubuntu.com/ubuntu trusty-security/restricted i386 Packages [17.8 kB]
Get:67 http://security.ubuntu.com/ubuntu trusty-security/restricted Translation-en [3,272 B]
Hit:68 http://ppa.launchpad.net/couchdb/stable/ubuntu trusty Release
Get:70 http://security.ubuntu.com/ubuntu trusty-security/universe amd64 Packages [297 kB]
Get:71 http://security.ubuntu.com/ubuntu trusty-security/universe i386 Packages [284 kB]
Get:72 http://ppa.launchpad.net/chris-lea/redis-server/ubuntu trusty/main amd64 Packages [1,706 B]
Get:73 http://security.ubuntu.com/ubuntu trusty-security/universe Translation-en [159 kB]
Get:74 http://ppa.launchpad.net/chris-lea/redis-server/ubuntu trusty/main i386 Packages [1,708 B]
Get:75 http://security.ubuntu.com/ubuntu trusty-security/multiverse amd64 Packages [4,718 B]
Get:76 http://security.ubuntu.com/ubuntu trusty-security/multiverse i386 Packages [4,877 B]
Get:77 http://ppa.launchpad.net/git-core/ppa/ubuntu trusty/main amd64 Packages [3,482 B]
Get:78 http://security.ubuntu.com/ubuntu trusty-security/multiverse Translation-en [2,426 B]
Get:79 http://ppa.launchpad.net/git-core/ppa/ubuntu trusty/main i386 Packages [3,482 B]
Get:80 http://ppa.launchpad.net/git-core/ppa/ubuntu trusty/main Translation-en [2,332 B]
Get:81 http://ppa.launchpad.net/openjdk-r/ppa/ubuntu trusty/main amd64 Packages [4,880 B]
Get:82 http://ppa.launchpad.net/openjdk-r/ppa/ubuntu trusty/main i386 Packages [4,884 B]
Get:83 http://ppa.launchpad.net/pollinate/ppa/ubuntu trusty/main amd64 Packages [430 B]
Get:84 http://ppa.launchpad.net/pollinate/ppa/ubuntu trusty/main i386 Packages [430 B]
Get:85 http://ppa.launchpad.net/pollinate/ppa/ubuntu trusty/main Translation-en [374 B]
Get:86 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty/main amd64 Packages [94.7 kB]
Get:87 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty/main i386 Packages [93.0 kB]
Get:88 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty/main Translation-en [21.0 kB]
Get:89 http://ppa.launchpad.net/webupd8team/java/ubuntu trusty/main amd64 Packages [1,549 B]
Get:90 http://ppa.launchpad.net/webupd8team/java/ubuntu trusty/main i386 Packages [1,549 B]
Get:91 http://ppa.launchpad.net/webupd8team/java/ubuntu trusty/main Translation-en [834 B]
Fetched 10.2 MB in 6s (1,636 kB/s)
Reading package lists...
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.4 Release: The following signatures were invalid: KEYEXPIRED 1515625755
W: GPG error: http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 1E9377A2BA9EF27F
W: The repository 'http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu trusty InRelease' is not signed.
W: There is no public key available for the following key IDs:
1E9377A2BA9EF27F  
W: http://ppa.launchpad.net/couchdb/stable/ubuntu/dists/trusty/Release.gpg: Signature by key 15866BAFD9BCC4F3C1E0DFC7D69548E1C17EAB57 uses weak digest algorithm (SHA1)
W: Failed to fetch http://repo.mongodb.org/apt/ubuntu/dists/trusty/mongodb-org/3.4/Release.gpg  The following signatures were invalid: KEYEXPIRED 1515625755
W: Some index files failed to download. They have been ignored, or old ones used instead.
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends --allow-downgrades --allow-remove-essential --allow-change-held-packages install ccache libssl1.0.0 bar time binutils gcc-5 g++-5 gcc-5-multilib g++-5-multilib make:i386 libssl-dev:i386 gfortran-5 gfortran-5-multilib" failed and exited with 100 during .
```

**Comment by mbesancon** [comment]
cc @JeffBezanson @StefanKarpinski if one of you can check if the description of `@irrational` makes sense :)

File: base/irrationals.jl [hunk 1]
```diff
@@ -136,6 +136,13 @@ end
 
 round(x::Irrational, r::RoundingMode) = round(float(x), r)
 
+"""
+	@irrational sym val def
+	@irrational(sym, val, def)
+
+	Provides a pre-computed value `val` for an irrational constant
+	identified by `sum`, along with its definition `def`.
```
[the remark this answers is not in the document]
Comment (mbesancon) [reply base/irrationals.jl]: thanks, updated

File: base/expr.jl [hunk 2]
```diff
@@ -199,6 +199,13 @@ macro noinline(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :noinline) : ex)
 end
 
+"""
+    @pure ex
+    @pure(ex)
+`@pure` gives the compiler a hint for the definition of a pure function, helping for type inference.
+A pure function only depends on its input which must be immutable (symbols, numbers)
+Use with caution, incorrect `@pure` annotation of a function may introduce bugs.
```
[the remark this answers is not in the document]
Comment (mbesancon) [reply base/expr.jl]: added some warning about generic functions and method tables

**Review by StefanKarpinski — COMMENTED** [review]
I still don’t think this explanation of `@pure` is correct. The arguments can be mutable but the result cannot depend on the mutable contents. Returning the `object_id` of an object should be pure, for example. The `===` function is also pure. Of course, there’s not that many useful functions of mutable objects that are pure but this explanation is still not quite right.

**Comment by mbesancon** [comment]
Ok this will be quicker to find the right doc directly from this thread:

---------------------------------------------
`@pure` gives the compiler a hint for the definition of a pure function,
helping for type inference.

A pure function can only depend on immutable information.
This also means a `@pure` function cannot use any global mutable state, including
generic functions. Calls to generic functions depend on method tables which are
mutable global state.
Use with caution, incorrect `@pure` annotation of a function may introduce
hard to identify bugs. Double check for calls to generic functions.

---------------------------------------------
Would that be ok?

**Comment by mbesancon** [comment]
just re-merged master, some CI tests seemed not to pass anymore

**Comment by mbesancon** [comment]
same thing, circleci didn't pass so I merged the latest master

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Thanks for the documentation improvements and sorry for the drawn out review process!

**Comment by matbesancon** [comment]
no trouble, there were more touchy and critical PRs on the stack :)

**Review by vtjnash — COMMENTED** [review]
whoops, github is telling me that apparently I had left some comments, but never submitted them 😬

**Comment by matbesancon** [comment]
@vtjnash I'll make another PR for these

File: base/int.jl [hunk 3]
```diff
@@ -553,14 +553,36 @@ floor(::Type{T}, x::Integer) where {T<:Integer} = convert(T, x)
...
     return parse(UInt128, s)
 end
 
+"""
+    @big_str str
+    @big_str(str)
+
+`@big_str` parses a string into a BigInt
+Throws an `ArgumentError` if the string is not a valid integer
+Removes all underscores `_` from the string
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply base/int.jl]: this has been solved on master

**Pull request merged**